### PR TITLE
Extract and restore MMR state

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- Added the ability to change the configuration for the MerkleChangeTracker, this allows the minimum and maximum checkpoint history to be set.
- Added functionality to allow the base MMR of the MerkleChangeTracker to be updated using the oldest checkpoints when the maximum checkpoint history is reached.
- Added the ability to retrieve the leaf hashes from MerkleMountainRange and restore it using leaf hashes.
- The state of MutableMmr and MerkleChangeTracker can be retrieved and used to restore their previous states.
- CompactMmrState was introduced to extract a restorable state for the MerkleChangeTracker and MutableMmr. It also allows state parts to be retrieved and combined to form the full state.
- Added clear to ArrayLike trait and shift to ArrayLikeExt and provided implementations for these functions for LMDBVec and for Vec<T>, this was required to better manage checkpoints and enable the restore functionality.

## Motivation and Context
These changes are required to enable sharing of MMR state between Base Nodes.

## How Has This Been Tested?
Tests were added to test the construction and restoring from MMR states and demonstrate the updating of the base MMR of the MerkleChangeTracker.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/mmr/src/mutable_mmr.rs [hunk 1]
```diff
@@ -98,6 +107,11 @@ where
         Ok((hash, deleted))
     }
 
+    /// Returns the number of leave nodes in the MMR.
+    pub fn get_leaf_count(&self) -> Result<usize, MerkleMountainRangeError> {
+        self.mmr.get_leaf_count()
```
Comment (CjS77) [inline base_layer/mmr/src/mutable_mmr.rs]: Isn't this `self.size`?

File: base_layer/mmr/src/change_tracker.rs [hunk 2]
```diff
@@ -138,15 +167,40 @@ where
         self.mmr.compress()
     }
 
+    // Check if the the number of checkpoints have exceeded the maximum configured history length. If it does then the
+    // oldest checkpoints are applied to the base mmr.
+    fn update_base_mmr(&mut self) -> Result<(), MerkleMountainRangeError> {
+        let checkpoint_count = self
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/mmr/src/change_tracker.rs]: Changed to make use of the member function.

File: base_layer/mmr/src/change_tracker.rs [hunk 3]
```diff
@@ -235,6 +289,11 @@ where
             .position(|h| h == hash)
             .map(|i| self.mmr.len() as usize - self.current_additions.len() + i)
     }
+
+    /// Returns the MMR state of the base MMR.
+    pub fn get_compact_state(&self, index: usize, count: usize) -> Result<CompactMmrState, MerkleMountainRangeError> {
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/mmr/src/change_tracker.rs]: Changed function name to `to_base_leaf_nodes `

File: base_layer/mmr/src/change_tracker.rs [hunk 4]
```diff
@@ -235,6 +289,11 @@ where
...
+        self.base.get_compact_state(index, count)
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/mmr/src/change_tracker.rs]: Changed function name to `to_leaf_nodes`

File: base_layer/mmr/src/merkle_mountain_range.rs [hunk 5]
```diff
@@ -79,11 +90,29 @@ where
             .map_err(|e| MerkleMountainRangeError::BackendError(e.to_string()))
     }
 
+    /// Returns the number of leave nodes in the MMR.
```
[the remark this answers is not in the document]
Comment (neonknight64) [reply base_layer/mmr/src/merkle_mountain_range.rs]: Change applied

File: base_layer/mmr/src/mutable_mmr.rs (hunk 1 above)
Comment (neonknight64) [reply base_layer/mmr/src/mutable_mmr.rs]: Correct, change applied.

**Pull request merged**